compiler/src/iree/compiler/Codegen/WGSL/WGSLReplacePushConstants.cpp

```diff
@@ -0,0 +1,137 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+// These must match what the runtime uses.
+#define IREE_HAL_WEBGPU_PARAMS_BIND_GROUP_INDEX 3
+#define IREE_HAL_WEBGPU_PARAMS_BINDING_INDEX 0
+
+static void replaceConstantLoadOp(IREE::Flow::DispatchTensorLoadOp loadOp,
+                                  IREE::HAL::InterfaceConstantLoadOp op) {
+  OpBuilder builder(op);
+
+  // tensor.extract -> i32
+  auto offsetValue = builder.createOrFold<arith::ConstantIndexOp>(
+      op.getLoc(), op.getIndex().getZExtValue());
+  auto extractOp =
+      builder.create<tensor::ExtractOp>(op.getLoc(), loadOp, offsetValue);
+
+  // Convert to original type.
+  auto opType = op.getType();
+  if (opType.isIndex()) {
+    auto indexCastOp =
+        builder.create<arith::IndexCastOp>(op.getLoc(), opType, extractOp);
+    op.replaceAllUsesWith(indexCastOp.getResult());
+  } else {
+    // TODO(scotttodd): truncate or zero-extend types with different bitwidths?
+    auto bitcastOp =
+        builder.create<arith::BitcastOp>(op.getLoc(), op.getType(), extractOp);
+    op.replaceAllUsesWith(bitcastOp.getResult());
+  }
+
+  op.erase();
+}
+
+class WGSLReplacePushConstantsPass
+    : public WGSLReplacePushConstantsBase<WGSLReplacePushConstantsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<mlir::arith::ArithmeticDialect, mlir::func::FuncDialect,
+                    mlir::tensor::TensorDialect, IREE::Flow::FlowDialect,
+                    IREE::HAL::HALDialect>();
+  }
+
+  void runOnOperation() override {
+    auto parentOp = getOperation();
```

Comment:
I am confused whats happening here. The `func.func` op should have the `IREE::HAL::InterfaceConstantLoadOp` in the body?
Also `parentOp` is confusing. Call it `funcOp`?